### ISSUE TITLE
plugin.classes.php - PHP Fatal error: Uncaught Error: [] operator not supported for strings

### ISSUE THREAD

**Issue by mrlilly** [jstanden/cerb#555]
A bug has been found in https://github.com/jstanden/cerb/blob/8.2.5/features/cerberusweb.core/api/plugin.classes.php
Lines 667 - 687

PHP message: PHP Fatal error:  Uncaught Error: [] operator not supported for strings in features/cerberusweb.core/api/plugin.classes.php:684

```
curl_setopt($ch, CURLOPT_HEADERFUNCTION, function($curl, $header) use (&$response_headers) {
			$len = strlen($header);
			
			$parts = explode(':', $header, 2);
			
			if(count($parts) < 2)
				return $len;
			
			$header_name = trim(strtolower($parts[0]));
			$header_value = trim($parts[1]);
			
			if(!isset($response_headers[$header_name])) {
				$response_headers[$header_name] = $header_value;
			} else {
				if(!isset($response_headers[$header_name]))
					$response_headers[$header_name] = [];
				
				$response_headers[$header_name][] = $header_value;
			}
			return $len;
		});
```

If the response has duplicate headers, you are setting the first as a string:
```
$response_headers[$header_name] = $header_value;
```
Then the next time trying to assign an array to a variable that was previously a string

```
$response_headers[$header_name][] = $header_value;
```
Not sure if something has been missed here as this part will never be executed as we are already in the else statement that checked the outcome of that a few lines up:
```
if(!isset($response_headers[$header_name]))
   $response_headers[$header_name] = [];
```

**Comment by jstanden** [comment]
Thanks for the report!

The first line of the `if` should be:

~~~
$response_headers[$header_name] = [$header_value];
~~~

I'll get that patched.

**Comment by jstanden** [comment]
Really, as you noted, the `if` doesn't even need to be there.  We should just store all header values as an array.

**Comment by jstanden** [comment]
This is what the function should be:

~~~
// See: https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2
curl_setopt($ch, CURLOPT_HEADERFUNCTION, function($curl, $header) use (&$response_headers) {
	$len = strlen($header);
	
	$parts = explode(':', $header, 2);
	
	if(count($parts) < 2)
		return $len;
	
	$header_name = trim(strtolower($parts[0]));
	$header_value = trim($parts[1]);
	
	if(!isset($response_headers[$header_name])) {
		$response_headers[$header_name] = $header_value;
	} else {
		$response_headers[$header_name] .= ', ' . $header_value;
	}
	
	return $len;
});
~~~

RFC-2616 says that any duplicate HTTP headers should be able to be condensed into a comma-separated list that preserves the order.  That way we can assume all the returned values are a string.

A duplicate HTTP response header, like `Cache-Control`, is the most likely scenario.

**Comment by jstanden** [comment]
Fixed in 8.2.6.  I'll push the fix in a few mins.

Thanks again.
